# Gearing Up for Deals — Analyzing Trends in Germany’s eBay Used Car Market

### Purpose of this Project
>This portfolio project is designed to run through an initial exploratory data analysis workflow, leveraging Python's `NumPy` and `pandas` libraries. A dataset containing 50,000 records is to be cleaned and prepared, followed by an initial analysis.


## Project Background
The project involves handling a dataset containing information on used cars from *eBay Kleinanzeigen*, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay marketplace. 

>The main aim of the project is to clean and prepare the dataset for analysis and perform some initial exploratory data analysis (EDA) on the prepared data. 

## About the Dataset
The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/datasets/sijovm/used-cars-data-from-ebay-kleinanzeigen). For this project, we are handling a sample of 50,000 records prepared by [Dataquest](https://www.dataquest.io) which have been 'dirtied' to simulate a more realistic version of a raw data file.

The following `data dictionary` describes each field contained within the original dataset:
>* `dateCrawled`: when the ad was first crawled (i.e., first recorded into the dataset). All field-values are taken from this date.
>
>* `name`: the complete name description of the vehicle.
>
>* `seller`: whether the seller is private or a dealer.
>
>* `offerType`: the type of listing ('offer' or 'wanted').
>
>* `price`: the price of the car listed on the ad.
>
>* `abtest`: whether the listing is included in an A/B test.
>
>* `vehicleType`: the type of vehicle.
>
>* `yearOfRegistration`: the year in which the vehicle was first registered.
>
>* `gearbox`: the vehicle's transmission type.
>
>* `powerPS`: the power of the car in horsepower (since dataset is in German, horsepower is known as **pferdestärke**).
>
>* `model`: the vehicle's model name.
>
>* `odometer`: the total number of kilometres the vehicle has driven as displayed on the odometer.
>
>* `monthOfRegistration`: the month in which the vehicle was first registered.
>
>* `fuelType`: the type of fuel the vehicle uses.
>
>* `brand`: the brand of vehicle.
>
>* `notRepairedDamage`: if the vehicle has damage that has not been repaired.
>
>* `dateCreated`: the date the eBay listing was created.
>
>* `nrOfPictures`: the number of pictures in the ad.
>
>* `postalCode`: the postal code for the location of the vehicle.
>
>* `lastSeenOnline`: when the crawler last saw this ad online.



### Loading the Dataset into a DataFrame

In [1]:
import numpy as np
import pandas as pd

The dataset is loaded, specifying `Latin-1` encoding since the raw CSV file uses this encoding (rather than UTF-8 which is standard).

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
autos.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

The `info()` method shows the dataframe contains 20 columns and 50,000 rows. Five of the fields store data using the `int64` datatype and 15 store data as the `object` type (likely text data). There are a handful of columns which contain null values, but no columns have more than ~20% null values.

Passing the argument `memory_usage="deep"` reveals the dataframe occupies 44MB of storage space. This is a **small** dataset - pandas can comfortably handle datasets of sizes up to around 3GB before modifications to the pandas workflow need to be considered.

The top 5 rows of the dataframe:

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Cleaning Column Labels

Reviewing the columns, it would be best to convert the column labels into the `snake_case` format rather than `camelCase` since this is convention for variable naming in Python. 

Some labels can also be shortened and clarified to improve readibility of code and the data.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code', 'last_seen']

In [7]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

### Initial Summary Stats and Cleaning

In [8]:
autos.describe() # includes only numeric columns

registration_year      power_ps  registration_month  num_photos  \
count       50000.000000  50000.000000        50000.000000     50000.0   
mean         2005.073280    116.355920            5.723360         0.0   
std           105.712813    209.216627            3.711984         0.0   
min          1000.000000      0.000000            0.000000         0.0   
25%          1999.000000     70.000000            3.000000         0.0   
50%          2003.000000    105.000000            6.000000         0.0   
75%          2008.000000    150.000000            9.000000         0.0   
max          9999.000000  17700.000000           12.000000         0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [9]:
autos.describe(include=["O"]) # includes only object columns

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   

       vehicle_type  gearbox  model   odometer fuel_type       brand  \
count         44905    47320  47242      50000     45518       50000   
unique            8        2    245         13         7          40   
top       limousine  manuell   golf  150,000km    benzin  volkswagen   
freq          12859    36993   4024      32424     30107       10687   

       unrepaired_damage           ad_created            last_seen  
count              40171                50000                50000  
unique                 2                   76                39481  
top                 nein  2016-04-03 00:00:00  2016-04-07 06:17:27  
freq               35232                 1946                    8

#### Reviewing the summary table above, there are a number of things to note:

---
**Columns which can be removed:**

* The `seller` column almost entirely contains the value 'privat' which indicates the listing is from a private seller rather than a dealership company. 

* The same applies to `offer_type` where nearly all values are the same ('Angebot', meaning the listing is an offer).

**Columns requiring additional investigation:**

* The `num_photos` column is odd since all values are zero.

* The extrema values in `registration_year` are nonsense (a vehicle cannot possibly be registered in the year 1000 or 9999).

* The extrema values in `power_ps` also require additional investigation as they seem unrealistic.

**Columns requiring a data type change:**
* `odometer` and `price` are stored in the `object` data type. These are much better suited to be stored as integers for sorting and calculations.


### Dropping Columns

In [10]:
autos["num_photos"].value_counts()

num_photos
0    50000
Name: count, dtype: int64

Since we have no way to know or infer what the `num_photos` values should be, it is best to drop this column entirely. We can also drop the `seller` and `offer_type` columns since these data are practically redundant. 

In [11]:
autos.drop(["num_photos", "seller", "offer_type"], axis=1, inplace=True)

### Casting `price` and `odometer` Columns to Numeric

In [12]:
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

Reviewing the unique values in the pre-cleaned `price` show prices are preceded by a `$` symbol and use commas as thousands separators. 

These can be stripped using the pandas str accessor to call the Python `replace` string method as a vectorized operation. The column can then be casted to an `int` dtype using the `.astype()` method.

Since we have removed the currency symbol, it is best to rename the column to reflect this information. This is done using the `.rename()` method.

In [13]:
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(int) 

autos.rename({"price":"price_usd"}, axis=1, inplace=True)

autos["price_usd"].head(10)

0    5000
1    8500
2    8990
3    4350
4    1350
5    7900
6     300
7    1990
8     250
9     590
Name: price_usd, dtype: int64

The same process is repeated for the `odometer` column:

In [14]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [15]:
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)

autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

autos["odometer_km"].head(10)

0    150000
1    150000
2     70000
3     70000
4    150000
5    150000
6    150000
7    150000
8    150000
9    150000
Name: odometer_km, dtype: int64

### Visualizing the Updated DataFrame

In [16]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price_usd  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0       5000  control          bus               2004    manuell       158   
1       8500  control    limousine               1997  automatik       286   
2       8990     test    limousine               2009    manuell       102   
3       4350  control   kleinwagen               2007  automatik        71   
4       1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [17]:
autos.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   price_usd           50000 non-null  int64 
 3   ab_test             50000 non-null  object
 4   vehicle_type        44905 non-null  object
 5   registration_year   50000 non-null  int64 
 6   gearbox             47320 non-null  object
 7   power_ps            50000 non-null  int64 
 8   model               47242 non-null  object
 9   odometer_km         50000 non-null  int64 
 10  registration_month  50000 non-null  int64 
 11  fuel_type           45518 non-null  object
 12  brand               50000 non-null  object
 13  unrepaired_damage   40171 non-null  object
 14  ad_created          50000 non-null  object
 15  postal_code         50000 non-null  int64 
 16  last_seen           50

*Note, we have reduced the memory usage by just over 10MB by removing redundant/unusable columns.*

Null values still need to be investigated further and the extreme values in `power_ps` and `registration_year` need to be addressed, but the dataset is in a good position to do some initial exploratory analysis on the `price_usd` and `odometer_km` columns.

### Exploring `odometer_km` and `price_usd`

In [18]:
autos["odometer_km"].value_counts()

odometer_km
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: count, dtype: int64

The `odometer_km` data values are all rounded to the nearest thousand. It is likely sellers had to select from a pre-filled drop-down list.

The distribution is skewed towards high mileage. ~79.5% of all vehicles in the dataset have mileages over 100,000 km.

In [19]:
autos["price_usd"].unique().shape

(2357,)

In [20]:
autos["price_usd"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

There are 2357 unique prices listed in the dataset. The minimum listed price is $0 USD whilst the maximum is $100 million USD. 

The minimum value is realistic considering the context as sellers may list old/unwanted or damaged cars to be claimed for free (as a means to dispose of junk). 

The maximum value seems highly irregular however.

In [21]:
autos["price_usd"].value_counts().head(20)

price_usd
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: count, dtype: int64

1421 listings in the dataset have a price of $0, which is the most frequent listed price. This makes up just under 2% of the entire dataset, but does suggest a notable portion of sellers may use the marketplace as a means to dispose of unwanted vehicles.

In [22]:
autos["price_usd"].value_counts().sort_index(ascending=False).head(20) # returns the top 20 listings with the highest prices, along with the number of listings with each price

price_usd
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: count, dtype: int64

Sorting by price in descending order reveals the frequency of listings with the largest prices. 

Looking at the prices, we see a steady price increase up to $350,000 USD and then a sudden jump up to $999,990 USD. Based on the context and this pattern in the data, it makes sense to discard the 14 records with prices exceeding $350,000 USD as these values are highly unrealistic (and likely are due to hoax/joke listings).

In [23]:
(autos["price_usd"]             # returns the top 20 listings with the lowest prices, along with the number of listings with each price
 .value_counts()
 .sort_index(ascending=True)
 .head(20)
 )

price_usd
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: count, dtype: int64

There are a number of listings with prices below $30 USD. Given eBay is primarily an auction site, it is reasonable to expect items to be featured with an opening bid as low as $1 USD.

For our price analysis, we should discard rows with prices exceeding $350,000 USD as well as $0 USD listings:

In [24]:
autos = autos[autos["price_usd"].between(1, 350000)]

autos["price_usd"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_usd, dtype: float64

We remove all rows from the `autos DataFrame` with prices falling outside of the 1-350,000 USD range. This is done by passing a Boolean mask (using the `.between()` method) to the dataframe and reassigning the result to the `autos` variable where the DataFrame is stored.

Using the `.desrcibe()` method on the `price_usd` column of the updated dataframe generates summary statistics for the cleaned `price_usd` column.

From the count vale, we can see we have removed 1435 listings: 

>* The resulting **median price** is `$3,000 USD`. 
>
>* The **IQR (Q3-Q1)** is `$6,290 USD` - the range of the central 50% of the data.

*Using the median and IQR in this case are more robust measures due to the wide spread of datapoints. The mean and std are far more sensitive to extrema, leading to less reliable measures of central tendency and spread respectively.*

### Cleaning the `unrepaired_damage` and `model` Columns

In [25]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   name                48565 non-null  object
 2   price_usd           48565 non-null  int64 
 3   ab_test             48565 non-null  object
 4   vehicle_type        43979 non-null  object
 5   registration_year   48565 non-null  int64 
 6   gearbox             46222 non-null  object
 7   power_ps            48565 non-null  int64 
 8   model               46107 non-null  object
 9   odometer_km         48565 non-null  int64 
 10  registration_month  48565 non-null  int64 
 11  fuel_type           44535 non-null  object
 12  brand               48565 non-null  object
 13  unrepaired_damage   39464 non-null  object
 14  ad_created          48565 non-null  object
 15  postal_code         48565 non-null  int64 
 16  last_seen           48565 n

The `unrepaired_damage` column contains 9,101 null values. These can be made more descriptive by replacing them with the string "not_specified".

In [26]:
autos.loc[autos["unrepaired_damage"].isnull(), "unrepaired_damage"] = "not_specified"

autos["unrepaired_damage"].value_counts(dropna=False)

unrepaired_damage
nein             34775
not_specified     9101
ja                4689
Name: count, dtype: int64

Filling null values in the `model` column is far more involved. Many of the null values within the model column should not be null values since the model of the vehicle is known from the `name` field in the record. 

We can partly address this below, filling the majority of the `model` null values with the correct vehicle model for the row (based on the `name` column). 

To do this, we first define the Series object `vehicle_names_with_null_models` which pulls the `name` column from the dataframe for every record containing a null value in the `model` column. Since the name column uses underscores to separate words, we can split the value of the name column about the underscore to generate an array of split names for every record.

In [27]:
vehicle_names_with_null_models = autos[autos["model"].isnull()]["name"]

split_names = vehicle_names_with_null_models.str.split("_")

split_names

10                      [VW, Golf, Tuning, in, siber/grau]
11                          [Mercedes, A140, Motorschaden]
15       [Corvette, C3, Coupe, T, Top, Crossfire, Injec...
23       [Peugeot, Boxer, 2, 2, HDi, 120, Ps, 9, Sitzer...
25       [Ford, escort, kombi, an, bastler, mit, ghia, ...
                               ...                        
49941                [Maserati, Ghibli, Diesel, Automatik]
49950    [Volvo, V70, 2, II, , 2002, , AHK, , Klima, , ...
49954                              [328, Cabrio, Schalter]
49985                       [Verkaufe, meinen, vw, vento!]
49993    [Audi, A3, , 1, 8l, , Silber;, schoenes, Fahrz...
Name: name, Length: 2458, dtype: object

We then assign a new variable `models` to store all unique vehicle models present in the dataset, casting this as a string datatype. We then append an array of models identified from the `name` field of vehicles which have null model values (see the `before_mapping` variable below to see how these models where determined).

The items in the `models` array are all made lowercase and then we loop over each row in the split_names series assigned in the code cell above. A variable `model_match` is assigned to "not_specified" as a default for every row. The inner loop then executes, updating the model_match variable to the item in the row array if it matches with one of the models in the `models` ndarray.

We then append all of the model_match strings to the mappings list. This is an ordered list, with each element corresponding to a model determined from the `name` field for that particular row. This `mappings` list can simply be assigned to the `model` column for all rows containing null values.

In [28]:
models = autos["model"].unique().astype(str)

models = np.append(models, ["vento", "a140", "ghibli", "h2", "dakota", "206", "626", "deville", "nitro", "cabrio", "challenger", "318", "spitfire", "tr3", "353", "explorer", "525", "grandis", "carens", "b12", "bonneville", "firebird", "c6", "gallardo", "demio", "530d", "330", "boxer", "caliber"])

models = np.char.lower(models)

mappings = []

for row in split_names:
  
  model_match = "not_specified"
  
  for item in row:
    
    if item.lower() in models:
      
      model_match = item
      
      break
     
  mappings.append(model_match)

In [29]:
autos.loc[autos["model"].isnull(), "model"] = "not_specified" # replaces all null values in the "model" column with "not_specified"

before_mapping = autos[autos["model"] == "not_specified"]["name"].value_counts().head(20) # returns the top 20 vehicle names with null values in the "model" column. we can clearly see model names in many of the vehicle names
before_mapping

name
Peugeot_206              9
Renault_Twingo           7
Dodge_RAM                6
Zu_verkaufen             5
MG_MGF_1.8i              5
Suche_ein_Auto           4
Hummer_H2                4
Mazda_626                4
Opel_Astra               4
Cadillac_Deville         3
Ford_mondeo              3
Opel_Vectra_B            3
Abarth_Grande_Punto      3
Renault_Megane_Scenic    3
Dodge_Nitro              3
Opel_vectra              3
Opel_corsa               3
Dodge_Challenger         2
Lincoln_Town_Car         2
Piaggio_Porter           2
Name: count, dtype: int64

In [30]:
print(mappings[:50]) # displays the first 50 elements in the mappings list which can be assigned to the model column to fill many of the missing values

print(len(mappings)) # the length of the mappings list indicates 2458 unspecified model values were present in the dataset. these are to be updated with the values in the mappings list

['Golf', 'A140', 'C3', 'Boxer', 'escort', 'Passat', 'not_specified', 'Vento', 'not_specified', 'not_specified', 'not_specified', 'not_specified', 'not_specified', 'STILO', 'not_specified', 'not_specified', 'not_specified', 'vectra', 'not_specified', 'Nitro', '206', 'Focus', 'not_specified', '601', 'not_specified', 'not_specified', 'A6', 'not_specified', 'C4', 'not_specified', 'not_specified', '200', 'not_specified', 'not_specified', 'not_specified', 'not_specified', 'not_specified', 'mondeo', 'Escort', 'omega', 'Firebird', 'Cabrio', 'sportage', 'Cherokee', 'not_specified', 'Polo', 'Polo', 'not_specified', 'Niva', 'not_specified']
2458


In [31]:
autos.loc[autos["model"] == "not_specified", "model"] = mappings # replaces all null values in the "model" column with the values in the mappings list

In [32]:
after_mapping = autos[autos["model"] == "not_specified"]["name"].value_counts().head(20) # returns the top 20 vehicle names with null values in the "model" column after mapping.
after_mapping

name
Dodge_RAM                                          6
MG_MGF_1.8i                                        5
Zu_verkaufen                                       5
Suche_ein_Auto                                     4
Ersatzteilspender_mit_Rest_TÜV_bis_Mai             2
Lexus_IS_250_Sport_Line                            2
MG_MGB                                             2
UNFAL_Auto                                         2
Mercedes_benz_ml_320_lpg                           2
Piaggio_Porter                                     2
Suche_Auto                                         2
Sommerreifen                                       2
Lincoln_Town_Car                                   2
MG_MG_TF_135                                       2
Auto_gesucht                                       2
Lexus_IS_220d_DPNR_Luxury_Line                     2
Wartburg_1.3                                       2
Aixam_City_Premium__LEDER__45km/h__ab_16_Jahren    1
Peugeot_106_1_4_55kw_Nuer_heute!!!!!!    

We see many of the remaining names with null `model` values after the mapping are more difficult to extract a model name from. 

Some listing names (e.g., Suche_ein_Auto) are not vehicle listings and are instead listings where people are expressing interest in finding a car:

In [33]:
autos.loc[autos["name"] == "Suche_ein_Auto",:]

date_crawled            name  price_usd ab_test vehicle_type  \
3894   2016-03-25 13:25:22  Suche_ein_Auto       2800    test          NaN   
5883   2016-03-28 12:56:23  Suche_ein_Auto          1    test          NaN   
41537  2016-03-18 23:47:25  Suche_ein_Auto        350    test          NaN   
43516  2016-03-16 12:37:15  Suche_ein_Auto       2000    test          NaN   

       registration_year gearbox  power_ps          model  odometer_km  \
3894                2005     NaN       150  not_specified       150000   
5883                2005     NaN         0  not_specified         5000   
41537               2000     NaN         0  not_specified         5000   
43516               2000     NaN         0  not_specified       150000   

       registration_month fuel_type           brand unrepaired_damage  \
3894                    0       NaN  sonstige_autos     not_specified   
5883                    0       NaN      volkswagen     not_specified   
41537                   0       NaN  sonstige_autos     not_specified   
43516                   0       NaN  sonstige_autos     not_specified   

                ad_created  postal_code            last_seen  
3894   2016-03-25 00:00:00        35641  2016-04-07 07:16:12  
5883   2016-03-28 00:00:00        61118  2016-04-02 16:48:00  
41537  2016-03-18 00:00:00        44359  2016-03-28 02:44:40  
43516  2016-03-16 00:00:00        78662  2016-04-06 06:16:23

In [34]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   name                48565 non-null  object
 2   price_usd           48565 non-null  int64 
 3   ab_test             48565 non-null  object
 4   vehicle_type        43979 non-null  object
 5   registration_year   48565 non-null  int64 
 6   gearbox             46222 non-null  object
 7   power_ps            48565 non-null  int64 
 8   model               48565 non-null  object
 9   odometer_km         48565 non-null  int64 
 10  registration_month  48565 non-null  int64 
 11  fuel_type           44535 non-null  object
 12  brand               48565 non-null  object
 13  unrepaired_damage   48565 non-null  object
 14  ad_created          48565 non-null  object
 15  postal_code         48565 non-null  int64 
 16  last_seen           48565 n

In [35]:
autos[autos["model"] == "not_specified"]["model"].value_counts() # after mapping, there are 1157 unspecified values in the "model" column

model
not_specified    1157
Name: count, dtype: int64

We have reduced now completely eliminated all null values in the model column (by replacing unknowns with not_specified) and filling in 1,301 null values with the actual model of the vehicle.

Depending on the future analysis goals, we would consider dropping these records, but for now we will leave them in place and move on to exploring other areas of the dataset.
___

### Filling Null Values in the Remaining Columns

Since it is impossible to accurately fill null values with the correct data for the `vehicle_type`, `gearbox` and `fuel_type` columns, we can fill these with "not_specified" to more accurately describe the status of the missing values. 

Depending on the course the analysis takes, records containing "not_specified" values could be dropped from the dataset if we are interested in drawing conclusions using these fields. For now, they can be left in place as we explore other areas of the dataset.

In [36]:
autos.loc[autos["vehicle_type"].isnull(), "vehicle_type"] = "not_specified"
autos.loc[autos["gearbox"].isnull(), "gearbox"] = "not_specified"
autos.loc[autos["fuel_type"].isnull(), "fuel_type"] = "not_specified"

autos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48565 non-null  object
 1   name                48565 non-null  object
 2   price_usd           48565 non-null  int64 
 3   ab_test             48565 non-null  object
 4   vehicle_type        48565 non-null  object
 5   registration_year   48565 non-null  int64 
 6   gearbox             48565 non-null  object
 7   power_ps            48565 non-null  int64 
 8   model               48565 non-null  object
 9   odometer_km         48565 non-null  int64 
 10  registration_month  48565 non-null  int64 
 11  fuel_type           48565 non-null  object
 12  brand               48565 non-null  object
 13  unrepaired_damage   48565 non-null  object
 14  ad_created          48565 non-null  object
 15  postal_code         48565 non-null  int64 
 16  last_seen           48565 n

Filling these null values has put the dataset into a state where there are no remaining null values. 

Cleaning is not yet complete since there are still unrealistic values within some fields that have not been investigated (such as the extreme values in the `registration_year` and `power_ps` columns highlighted in the section above). These will be addressed in due course as we explore the dataset further.

## Exploring the Date Columns

**There are a handful of columns with date information, two are meta-data from the crawler and the other three are dates and times that were part of each listing:**

* `date_crawled` - the date and time when the web crawler recorded the listing into the dataset.
* `last_seen` - the date and time when the web crawler last saw the listing on eBay.
---
* `ad_created` - the date and time when the listing was posted as recorded by eBay.
* `registration_month` - the month when the vehicle was first registered for the road, recorded on the eBay listing.
* `registration_year` - the year when the vehicle was first registered for the road, recorded on the eBay listing.

The `date_crawled`, `last_seen` and `ad_created` columns are all defined as objects in the pandas df. These will need to be converted to numeric to understand them quantitatively.

In [37]:
autos[["date_crawled", "last_seen", "ad_created"]][:5]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

The first ten characters for each field represent the date. We can extract just the date values and use the `.value_counts()` method to generate a distribution.

### `date_crawled` Column

In [38]:
(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index(ascending=True)
 )

date_crawled
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: proportion, dtype: float64

In [39]:
(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_values(ascending=True)
 )

date_crawled
2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: proportion, dtype: float64

We see the listings were crawled over a period of around one month (5th March - 7th April 2016). The proprotion of listings crawled each day remains fairly consistent: **around 3% of the total listings were recorded each day**. The distribution is uniform. 

We do note an **apparent drop in listings recorded on the final three days** (down to around 1% on the 5th and then less than 0.5% on the 6th and 7th of April). This is **likely due to the crawling period ending** (presumably the crawler was instructed to scrape the site for a month) rather than a decline in the number of new listings being posted to the site.

### `last_seen` Column

In [40]:
(autos["last_seen"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index(ascending=True)
 )

last_seen
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: proportion, dtype: float64

The crawler recorded the last date it saw any of the listings in the dataset, which acts as a proxy to determine the day at which a vehicle listing was removed from the site (presumably because it was sold).

Similar to the trend in the `date_crawled` column, we see a disproportionate amount of "last seen" values on the final three listed days. Again, this is likely due to the crawling period ending since the values are abnormally high (between 6-10x the values for the previous days) rather than an indicator of a huge sales spike.

### `ad_created` Column

In [41]:
(autos["ad_created"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index(ascending=True)
 .head(20)
 )

ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
Name: proportion, dtype: float64

In [42]:
(autos["ad_created"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index(ascending=True)
 .tail(30)
 )

ad_created
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21    0.037579
2016-03-22    0.032801
2016-03-23    0.032060
2016-03-24    0.029280
2016-03-25    0.031751
2016-03-26    0.032266
2016-03-27    0.030989
2016-03-28    0.034984
2016-03-29    0.034037
2016-03-30    0.033501
2016-03-31    0.031875
2016-04-01    0.033687
2016-04-02    0.035149
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: proportion, dtype: float64

In [43]:
(autos["ad_created"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .shape
 )

(76,)

There are 76 distinct "ad created" dates in the dataset, with most falling within 1-2 months of the crawling period. There are some older listings featured in the dataset, with the oldest (11th June 2015) being around 9 months outside of the crawling period.

Looking at the oldest listing in the dataset, we see the vehicle on offer is a Mercedes Benz C220, with a listed price of $47,900 USD. This is clearly not a budget option, which may explain why the vehicle has not been sold after 9 months of being featured on the site.

In [44]:
autos[autos["ad_created"].str[:10] == "2015-06-11"]

date_crawled                                               name  \
22781  2016-03-23 01:48:59  Mercedes_Benz_C220_BT__7G_Tronic_AMG__Modellja...   

       price_usd ab_test vehicle_type  registration_year    gearbox  power_ps  \
22781      47900    test    limousine               2014  automatik         0   

          model  odometer_km  registration_month fuel_type          brand  \
22781  c_klasse        20000                   2    diesel  mercedes_benz   

      unrepaired_damage           ad_created  postal_code            last_seen  
22781     not_specified  2015-06-11 00:00:00        46145  2016-04-06 09:47:02

### `registration_year` Column

In [45]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year a vehicle was first registered will likely indicate the age of the car. 

As mentioned previously, we have highly abnormal extrema values (`1000` and `9999`) which will need handling before drawing conclusions from our initial descriptive analysis.

## Handling Inaccurate Registration Year Data

Vehicles that have registration years greater than 2016 are certainly inaccurate as all records from the dataset were collected in the first few months of 2016. 

Determining the earliest valid registration year is more difficult. Realistically, a vehicle could have been first registered as far back as the first few decades of the 1900s, so we can use 1900 as a conservative lower bound for the earliest valid registration year.

To begin, we can identify records with registration years outside of the range 1900-2016 and drop these rows.

In [46]:
autos["registration_year"].value_counts().sort_index(ascending=True).tail(13) # returns all unique registration year values in the dataset that exceed the upper limit for valid year

registration_year
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: count, dtype: int64

In [47]:
(~autos["registration_year"].between(1000, 2016)).sum() / autos.shape[0] # returns the proportion of all data points that are above the upper limit for valid registration year

np.float64(0.03869041490785545)

There are 13 unique registration year values above the upper bound for valid registration years. This corresponds to ~3.9% of all data in the dataset.

In [48]:
autos["registration_year"].value_counts().sort_index(ascending=False).tail()

registration_year
1910    5
1800    2
1111    1
1001    1
1000    1
Name: count, dtype: int64

Repeating this for the lower limit, there are five total records with registration years falling below 1900. 

Since the combined percentage of invalid values in this column is **less than 4%**, we can safely **drop these rows** without affecting our sample size too significantly:

In [49]:
autos = autos[autos["registration_year"].between(1900, 2016)]

In [50]:
autos["registration_year"].value_counts().sort_index(ascending=True)

registration_year
1910       5
1927       1
1929       1
1931       1
1934       2
        ... 
2012    1310
2013     803
2014     663
2015     392
2016    1220
Name: count, Length: 78, dtype: int64

The earliest registration year that falls within the 1900-2016 year range is 1910 with five records containing this value.

We can retrieve these records to check whether the year is accurate:

In [51]:
autos[autos["registration_year"] == 1910]["name"]

3679                           Suche_Auto
22659                        Opel_Corsa_B
28693                      Renault_Twingo
30781    Opel_Calibra_V6_DTM_Bausatz_1:24
45157                          Motorhaube
Name: name, dtype: object

In [52]:
autos[(autos["registration_year"] == 1910) & ((autos["name"] == "Suche_Auto") | (autos["name"] == "Motorhaube"))]

date_crawled        name  price_usd  ab_test   vehicle_type  \
3679   2016-04-04 00:36:17  Suche_Auto          1     test  not_specified   
45157  2016-03-11 22:37:01  Motorhaube         15  control  not_specified   

       registration_year        gearbox  power_ps          model  odometer_km  \
3679                1910  not_specified         0  not_specified         5000   
45157               1910  not_specified         0  not_specified         5000   

       registration_month      fuel_type           brand unrepaired_damage  \
3679                    0  not_specified  sonstige_autos     not_specified   
45157                   0  not_specified         trabant     not_specified   

                ad_created  postal_code            last_seen  
3679   2016-04-04 00:00:00        40239  2016-04-04 07:49:15  
45157  2016-03-11 00:00:00        90491  2016-03-25 11:18:57

* From these listings names, the three vehicles with row indexes `22659`, `28693` and `30781` are far more modern than 1910.

* The "Suche_Auto" listing is not a listing for an actual vehicle, instead an expression of interest in aquiring a car. It is unclear why the registration year was filled as 1910 by the poster in this case.

* The "Motorhaube" listing is similar to the "Suche_Auto" listing - it seems to be someone expressing a desire to purchase a car. In this case, the desired car brand is listed as a "trabant" which is a much older car brand, though these were manufacturered in 1957 at the earliest. 

**Therefore, all of these registration years are inaccurate too.** We can drop these rows from the dataset as well:

In [53]:
autos = autos[autos["registration_year"].between(1911, 2016)]

In [54]:
autos["registration_year"].value_counts().sort_index(ascending=True).head(10)

registration_year
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
Name: count, dtype: int64

In [55]:
autos[autos["registration_year"].between(1927, 1948)][["registration_year", "name"]]

registration_year                                               name
2221                1934  Sehr_seltener_Oldtimer_Opel_1210_zum_Restaurieren
2573                1934              Hanomag_rekord_15k_Suche_ersatz_teile
11047               1948          Andere_Simca_5_Fourgonette_Kombilimousine
11246               1931                  Ford_Model_A_Roadster_Deluxe_1931
11585               1943                              Volkswagen__VW_Typ_82
13963               1941  Mercedes_Benz_L1500S_Wehrmacht_/_Luftwaffe___F...
21416               1927                            Essex_super_six__Ford_A
21421               1937            Ford_Business_Coupe_Hotrod_Projekt.1937
22101               1929                                         BMW_Andere
23804               1937                                 Mercedes_Benz_170V
24855               1939                      Andere_DKW_F8_Cabrio_Limosine
25792               1941              Andere_GAZ_Oldtimergelaendewagen_1941
26103               1938               Oldtimer_Sport___Cabrio_Georges_Irat
26607               1937                                          MG_Andere
39725               1937  Dodge_1937_Hotrod_Flathead_Scheunenfund_Oldtim...

Briefly reviewing the remaining earliest registration years, we see these listings are for older cars manufactured around the timeframe listed in the registration year field. Based on the name field for the listing, these registration years appear to be accurate.  

### Distribution of the Cleaned `registration_year` Column

In [56]:
autos["registration_year"].value_counts(normalize=True).head(10)

registration_year
2000    0.067615
2005    0.062902
1999    0.062066
2004    0.057910
2003    0.057824
2006    0.057203
2001    0.056474
2002    0.053261
1998    0.050626
2007    0.048783
Name: proportion, dtype: float64

In [57]:
autos["registration_year"].value_counts(normalize=True).head(10).sum()

np.float64(0.5746636387008313)

The majority of the vehicles in the dataset were first registered in the past 25 years. The top 10 registration years featured in the dataset fall within the range 1998-2007, comprising of 57% of all vehicle listings in the dataset.

In [58]:
autos["registration_year"].describe()

count    46676.000000
mean      2002.920709
std          7.120843
min       1927.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

We see the mean and median of the registration year column are approximately the same, `2003`. 

The mean is slightly lower than the median, indicating a left skewed distribution. This makes sense in the context of the data - there are few listings with old registration years (<1970), the majority fall within the years 1998-2007.

### Exploring the `power_ps` Column

After handling the registration year column, we have dropped another chunk of records. We can further clean the dataset by addressing the extreme values in the `power_ps` (horsepower) column.

In [59]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 46676 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        46676 non-null  object
 1   name                46676 non-null  object
 2   price_usd           46676 non-null  int64 
 3   ab_test             46676 non-null  object
 4   vehicle_type        46676 non-null  object
 5   registration_year   46676 non-null  int64 
 6   gearbox             46676 non-null  object
 7   power_ps            46676 non-null  int64 
 8   model               46676 non-null  object
 9   odometer_km         46676 non-null  int64 
 10  registration_month  46676 non-null  int64 
 11  fuel_type           46676 non-null  object
 12  brand               46676 non-null  object
 13  unrepaired_damage   46676 non-null  object
 14  ad_created          46676 non-null  object
 15  postal_code         46676 non-null  int64 
 16  last_seen           46676 n

In [60]:
autos["power_ps"].describe()

count    46676.000000
mean       117.905562
std        184.928789
min          0.000000
25%         75.000000
50%        109.000000
75%        150.000000
max      17700.000000
Name: power_ps, dtype: float64

We can set upper and lower bounds for acceptable horsepower values. 

* Generally, top of the range supercars can have up to 900 break-horsepower, so this is a good starting point as an upper bound. 
* Vehicles with the lowest horsepower are typically micro-cars which can have as little as 4 break-horsepower.

Before making a decision on the range, we should first assess the accuracy of each of these horsepower values.

In [61]:
autos["power_ps"].value_counts().sort_index(ascending=True).head(20)

power_ps
0     4502
1        4
2        2
3        2
4        4
5       13
6        3
8        2
9        1
10       2
11       4
14       1
15       5
16       1
18       6
19       2
20       4
21       1
23       3
24       1
Name: count, dtype: int64

There are a disproportionately large number of listings with zero horsepower. This should be investigated:

In [62]:
autos[autos["power_ps"] == 0].head(10)

date_crawled                                               name  \
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
8   2016-03-22 16:51:34                                         Seat_Arosa   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
32  2016-03-20 05:03:03                               Corsa_mit_TÜV_5.2016   
35  2016-03-29 21:52:56                                        Ford_Mondeo   
46  2016-03-31 10:53:28                                       BMW_mit__Lpg   
52  2016-03-25 18:50:03  Senator_A_3.0E_Karosserie_restauriert_m._viele...   
59  2016-03-17 17:50:54               Mercedes_A_Klasse_W_168__A_140_gruen   
64  2016-04-05 07:36:19                 Autotransport__Abschlepp_Schlepper   
74  2016-03-06 02:00:38                                         Opel_Kombi   

    price_usd  ab_test   vehicle_type  registration_year        gearbox  \
4        1350     test          kombi               2003        manuell   
8         250     test  not_specified               2000        manuell   
11        350  control  not_specified               2000  not_specified   
32        350  control     kleinwagen               1999        manuell   
35       1200     test          kombi               1998        manuell   
46       9000  control  not_specified               2005      automatik   
52       3500     test      limousine               1985  not_specified   
59        700  control  not_specified               2016        manuell   
64         40     test  not_specified               2011  not_specified   
74       1500  control          kombi               2006        manuell   

    power_ps          model  odometer_km  registration_month      fuel_type  \
4          0          focus       150000                   7         benzin   
8          0          arosa       150000                  10  not_specified   
11         0           A140       150000                   0         benzin   
32         0          corsa       150000                   7         benzin   
35         0         mondeo       150000                  12         benzin   
46         0  not_specified       150000                   0  not_specified   
52         0         andere         5000                   0         benzin   
59         0       a_klasse       150000                   0         benzin   
64         0            5er       150000                   5  not_specified   
74         0          astra       150000                   3         benzin   

            brand unrepaired_damage           ad_created  postal_code  \
4            ford              nein  2016-04-01 00:00:00        39218   
8            seat              nein  2016-03-22 00:00:00         7426   
11  mercedes_benz     not_specified  2016-03-16 00:00:00        17498   
32           opel     not_specified  2016-03-20 00:00:00        27619   
35           ford              nein  2016-03-29 00:00:00        52428   
46            bmw     not_specified  2016-03-31 00:00:00        12351   
52           opel              nein  2016-03-25 00:00:00        63500   
59  mercedes_benz     not_specified  2016-03-17 00:00:00        95356   
64            bmw     not_specified  2016-04-05 00:00:00        40591   
74           opel              nein  2016-03-06 00:00:00        23627   

              last_seen  
4   2016-04-01 14:38:50  
8   2016-03-26 18:18:10  
11  2016-03-16 18:45:34  
32  2016-04-06 03:15:20  
35  2016-04-06 08:45:35  
46  2016-04-06 03:44:41  
52  2016-04-07 00:46:00  
59  2016-03-19 17:46:47  
64  2016-04-07 12:16:01  
74  2016-04-06 01:46:40

It appears records with a `power_ps` value of 0 are likely to be placeholder values where the crawler could not retrieve an accurate horsepower for a listing (this may be because the seller did not provide this information or if the crawler could not identify the value).

Since 4,502 records have this value, it would be better to fill these with a null value rather than dropping these rows entirely (as this is almost 10% of the remaining data). 

In [63]:
autos.loc[autos["power_ps"] == 0, "power_ps"] = np.nan

In [64]:
autos["power_ps"].value_counts().sort_index(ascending=True).head(20)

power_ps
1.0      4
2.0      2
3.0      2
4.0      4
5.0     13
6.0      3
8.0      2
9.0      1
10.0     2
11.0     4
14.0     1
15.0     5
16.0     1
18.0     6
19.0     2
20.0     4
21.0     1
23.0     3
24.0     1
25.0     2
Name: count, dtype: int64

In [65]:
autos["power_ps"].value_counts().sort_index(ascending=True).tail(20)

power_ps
1781.0     1
1793.0     1
1796.0     1
1800.0     1
1988.0     1
1998.0     1
2729.0     1
3500.0     1
3750.0     1
4400.0     1
5867.0     1
6045.0     1
6226.0     1
6512.0     1
7511.0     1
8404.0     1
14009.0    1
15001.0    1
16312.0    1
17700.0    1
Name: count, dtype: int64

There are a number of listings with very low or extremely large horsepower values which are very likely to be inaccurate (refer to the four records with a `power_ps` value of 1 below). 

In [66]:
autos[autos["power_ps"] == 1].head()

date_crawled                                               name  \
485    2016-03-22 17:57:58                            Super_Chrysler_neon_2.0   
844    2016-03-22 13:38:51              Golf_Cabrio_zum_Schlachten_ohne_Motor   
32820  2016-03-08 11:38:54  **_Golf_3**_in_Klassikrot_LA3G_91_98__3_Tuerig...   
46817  2016-03-07 19:53:06  Opel_Kadett_C_Scheunenfund_1._Serie_Trommelbre...   

       price_usd  ab_test vehicle_type  registration_year  gearbox  power_ps  \
485          490  control    limousine               2002  manuell       1.0   
844          300  control       cabrio               1997  manuell       1.0   
32820        200  control    limousine               1992  manuell       1.0   
46817        990  control        coupe               1974  manuell       1.0   

        model  odometer_km  registration_month      fuel_type       brand  \
485    andere       100000                  11  not_specified    chrysler   
844      golf         5000                   4         diesel  volkswagen   
32820    golf       100000                  12         benzin  volkswagen   
46817  kadett         5000                   3         benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
485                nein  2016-03-22 00:00:00        40233  2016-03-22 17:57:58  
844                  ja  2016-03-22 00:00:00        83395  2016-04-04 01:15:36  
32820                ja  2016-03-08 00:00:00        47506  2016-03-10 02:46:26  
46817                ja  2016-03-07 00:00:00        94036  2016-03-10 04:44:22

To summarise the horsepower data, we can exclude the anomalously large/small values from the calculation without the need to drop all these records from the dataset.

This is a good compromise to better assess the status of the accurate horsepower values whilst also keeping data for the records with unreliable horsepower values (for other analyses that do not require accurate `power_ps` values).

To ensure the horsepower values are most likely reliable, we can restrict the range in our summary calculation to 70-700. This covers the typical range of many modern vehicles from small cars up to heavy-duty trucks.

In [67]:
autos["power_ps"].between(70, 700).value_counts()

power_ps
True     35574
False    11102
Name: count, dtype: int64

Limiting the values to this range excludes 11,102 records from the calculation, though this is warranted considering the unreliable nature of this field.

In [68]:
autos.loc[autos["power_ps"].between(70, 700), "power_ps"].describe()

count    35574.000000
mean       139.340136
std         59.430756
min         70.000000
25%        101.000000
50%        125.000000
75%        163.000000
max        696.000000
Name: power_ps, dtype: float64

From the ~35k records, the `median horsepower is 125` - right around the range of typical mid-sized family cars. 

The `mean of 139` is notably larger than the median value indicating the distribution is right skewed (i.e., a greater number of lower horsepower vehicles are listed on the site).

The spread of the data is relatively compact, with an `IQR of 62`. This means the central 50% of the datapoint differ over a range of 62 horsepower values.

### Exploring Price by Vehicle Brand

In [69]:
brand_proportions = autos["brand"].value_counts(normalize=True)

brand_proportions

brand
volkswagen        0.211286
bmw               0.110057
opel              0.107550
mercedes_benz     0.096474
audi              0.086576
ford              0.069907
renault           0.047133
peugeot           0.029844
fiat              0.025645
seat              0.018275
skoda             0.016411
nissan            0.015276
mazda             0.015190
smart             0.014161
citroen           0.014011
toyota            0.012705
hyundai           0.010027
sonstige_autos    0.009791
volvo             0.009148
mini              0.008763
mitsubishi        0.008227
honda             0.007841
kia               0.007070
alfa_romeo        0.006642
porsche           0.006127
suzuki            0.005935
chevrolet         0.005699
chrysler          0.003514
dacia             0.002635
daihatsu          0.002507
jeep              0.002271
subaru            0.002142
land_rover        0.002100
saab              0.001650
jaguar            0.001564
daewoo            0.001500
trabant           0.00

Looking at the breakdown of brands in the dataset, German manufacturers make up four of the top five brands with almost 50% of the total listings from German brands. Volkswagen is the most popular brand by a large margin, making up 21.1% of all listings.

Since there are many brands that are less common, we can just take a look at the most common brands - those which make up more than 5% of total listings:

In [70]:
common_brands = brand_proportions[brand_proportions > 0.05].index

common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object', name='brand')

In [71]:
price_by_brand = {}

for brand in common_brands:
  
  selected_rows = autos[autos["brand"] == brand]
  
  mean_price_usd = selected_rows["price_usd"].mean()
  
  price_by_brand[brand] = f"{mean_price_usd:.0f}"
  
price_by_brand

{'volkswagen': '5402',
 'bmw': '8333',
 'opel': '2976',
 'mercedes_benz': '8628',
 'audi': '9337',
 'ford': '3749'}

In [72]:
price_by_brand = pd.Series(price_by_brand)

brand_aggregate_data = pd.DataFrame(price_by_brand, columns=["mean_price_usd"])

brand_aggregate_data

mean_price_usd
volkswagen              5402
bmw                     8333
opel                    2976
mercedes_benz           8628
audi                    9337
ford                    3749

Of the most common brands:

>* BMW, Mercedes and Audi are much more expensive options.
>
>* Opel and Ford are more affordable choices.
>
>* Volkswagen is intermediary, offering a balance between higher-end quality and price. This (along with the fact that VW is a German manufacturer) could explain the high popularity.

### Exploring Mileage by Brand

In [73]:
mileage_by_brand = {}

for brand in common_brands:
  
  selected_rows = autos[autos["brand"] == brand]
  
  mean_mileage = (selected_rows["odometer_km"] * 0.621371).mean() # convert km to miles
  
  mileage_by_brand[brand] = f"{mean_mileage:.0f}"
  
mileage_by_brand

{'volkswagen': '79975',
 'bmw': '82377',
 'opel': '80351',
 'mercedes_benz': '81268',
 'audi': '80255',
 'ford': '77215'}

In [74]:
mileage_by_brand = pd.Series(mileage_by_brand)

brand_aggregate_data["mean_mileage"] = mileage_by_brand

brand_aggregate_data.sort_values(by="mean_price_usd", ascending=False)

mean_price_usd mean_mileage
audi                    9337        80255
mercedes_benz           8628        81268
bmw                     8333        82377
volkswagen              5402        79975
ford                    3749        77215
opel                    2976        80351

Mean mileage by brand does not vary to the extent that mean price does. By simple inspection, there does seem to be a very slight tendency for less expensive brands to have a lower mileage on average.

### Conclusions

The `autos.csv` dataset has been cleaned and some initial exploratory data analysis has been conducted to gain insight into the nature of the dataset. 

* Through the cleaning process, an understanding of the date and horsepower fields has been developed, and many of the missing/inaccurate values have been handled ready for further data analysis.

* Some initial analysis was conducted on the price and mileage of vehicles, aggregating by brand to identify trends.

* Depending on the direction of the analysis, further cleaning may be necessary. If we wanted to explore potential trends regarding the vehicle horsepower, further cleaning of the `power_ps` field would be required. 

In [75]:
autos.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
Index: 46676 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   date_crawled        46676 non-null  object 
 1   name                46676 non-null  object 
 2   price_usd           46676 non-null  int64  
 3   ab_test             46676 non-null  object 
 4   vehicle_type        46676 non-null  object 
 5   registration_year   46676 non-null  int64  
 6   gearbox             46676 non-null  object 
 7   power_ps            42174 non-null  float64
 8   model               46676 non-null  object 
 9   odometer_km         46676 non-null  int64  
 10  registration_month  46676 non-null  int64  
 11  fuel_type           46676 non-null  object 
 12  brand               46676 non-null  object 
 13  unrepaired_damage   46676 non-null  object 
 14  ad_created          46676 non-null  object 
 15  postal_code         46676 non-null  int64  
 16  last_seen

In [76]:
autos.to_csv("autos_cleaned_v1.csv", index=False)

To end the cleaning process, it is best practice to save a copy of the cleaned dataset as a separate file format. The index parameter is set to false since our data already has numeric index labels, so these do not need to be added to the csv file.